---
title: "Untitled"
author: "Briha Ansari"
date: "6/21/2021"
output: html_document
---


```{r}
library(curatedTCGAData)
library(MultiAssayExperiment)
library(TCGAutils)
library(tidyverse)
#BiocManager::install("TCGAutils")
```


check disease codes in curatedTCGA data
```{r}
# head(
#   curatedTCGAData(
#     diseaseCode = "PRAD", assays = "*", version = "2.0.1"
#   )
# )
# 
# 
# data('diseaseCodes', package = "TCGAutils")
# head(diseaseCodes)
# diseaseCodes


```


check samples: Below is a glossary of each sample type and its meaning

SummarizedExperiment*
RNASeqGene
RSEM TPM gene expression values
RNASeq2GeneNorm
#Upper quartile normalized RSEM TPM geneexpression values
miRNAArray
Probe-level miRNA expression values
#miRNASeqGene
Gene-level log2 RPM miRNA expressionvalues
mRNAArray
Unified gene-level mRNA expression values
mRNAArray_huex
Gene-level mRNA expression values fromAffymetrix Human Exon Array
mRNAArray_TX_g4502a
Gene-level mRNA expression values fromAgilent 244K Array
mRNAArray_TX_ht_hg_u133a
Gene-level mRNA expression values fromAffymetrix Human Genome U133 Array
GISTIC_AllByGene
Gene-level GISTIC2 copy number values
GISTIC_ThresholdedByGene
Gene-level GISTIC2 thresholded discretecopy number values
#RPPAArray
Reverse Phase Protein Array normalizedprotein expression values


RangedSummarizedExperiment
GISTIC_Peaks
GISTIC2 thresholded discrete copy numbervalues in recurrent peak regions
SummarizedExperiment with HDF5Array DelayedMatrix
Methylation_methyl27
Probe-level methylation beta values fromIllumina HumanMethylation 27K BeadChip
Methylation_methyl450
Probe-level methylation beta values fromInfinium HumanMethylation 450K BeadChip


RaggedExperiment
CNASNP
Segmented somatic Copy NumberAlteration calls from SNP array
CNVSNP
Segmented germline Copy Number Variantcalls from SNP Array
CNASeq
Segmented somatic Copy NumberAlteration calls from low pass DNASequencing
#Mutation*
Somatic mutations calls
CNACGH_CGH_hg_244a
Segmented somatic Copy NumberAlteration calls from CGH Agilent Microarray244A
CNACGH_CGH_hg_415k_g4124a
Segmented somatic Copy NumberAlteration calls from CGH Agilent Microarray415K
All can be converted to RangedSummarizedExperiment (except RPPAArray) withTCGAutils
```{r}
# head(
#   curatedTCGAData(
#     diseaseCode = "PRAD", assays = "*", version = "2.0.1"
#   )
# )
# 
# 
   curatedTCGAData(
    diseaseCode = "PRAD", assays = "*", version = "2.0.1"   )
```

keep assays of interest
```{r}
prad <- suppressMessages(curatedTCGAData("PRAD",
                                         c("RPPAArray","miRNASeqGene","RNASeq2GeneNorm","Mutation"),
                                         dry.run=FALSE,version="2.0.1"))  
  
```

```{r}
prad


names(assays(prad))

```


Now we have the following assays in patients with Prostate cancer: 

1.RNASeq2GeneNorm: Upper quartile normalized RSEM TPM geneexpression values
2.miRNASeqGene:Gene-level log2 RPM miRNA expressionvalues
3.RPPAArray:Reverse Phase Protein Array normalizedprotein expression values
4.Mutation:Somatic mutations calls



check the type samples by sampletables()
Below are the codes for each sample type

#> Code Definition Short.Letter.Code
#> 1 01 Primary Solid Tumor TP
#> 2 02 Recurrent Solid Tumor TR
#> 3 03 Primary Blood Derived Cancer - Peripheral Blood TB
#> 4 04 Recurrent Blood Derived Cancer - Bone Marrow TRBM
#> 5 05 Additional - New Primary TAP
#> 6 06 Metastatic TM
```{r}

sampleTables(prad)
```
01 








keeping only primary tumors
```{r}
prad <- TCGAprimaryTumors(prad)

#recheck
sampleTables(prad)

```
Creating objects for each type of assay in our dataset
```{r}
mir <- assays(prad)[["PRAD_miRNASeqGene-20160128"]]
head(mir[,1:5])

methyl <- assays(prad)[["PRAD_RPPAArray-20160128"]]
head(methyl[,1:5])

mut <- assays(prad)[["PRAD_Mutation-20160128"]]
head(mut[,1:20])

rna <- assays(prad)[["PRAD_RNASeq2GeneNorm-20160128"]]
head(rna[,1:20])

```



Now explore clinical data

```{r}
pheno1 <- colData(prad)
```




```{r}
dim(pheno1)

```
check each variable of interest 
```{r}
names(pheno1)


#recurrence are colnames 216,186,159
table(pheno1[,216])
table(pheno1[,186])
table(pheno1[,159])

# # progression 193,164
# table(pheno1[,193])
# table(pheno1[,164])
# 
# # primary therapy outcome success 161
# table(pheno1[,161])


```


We will keep data with documented recurrence using column [186] "patient.follow_ups.follow_up.new_tumor_event_after_initial_treatment"

```{r}

staget <- sub("[abcd]","",sub("t","",colData(prad)$pathology_T_stage))
staget <- suppressWarnings(as.integer(staget))
colData(prad)$STAGE <- staget
tmp <- colData(prad)$patient.follow_ups.follow_up.new_tumor_event_after_initial_treatment 
str(tmp)

#create iDX vector for responses

idx <-which(tmp %in% c("yes","no"))
stagetna<-  which(is.na(staget))


length(idx)
length(stagetna)
pID <- colData(prad)$patientID

# keep IDs that have yes or no for recurrence - NAs for stage(T)
tokeep <- setdiff (pID[idx], pID[stagetna]) 
length(tokeep)
prad <- prad[,tokeep,]
pam50 <- colData(prad)$patient.follow_ups.follow_up.new_tumor_event_after_initial_treatment 


### where a patient has multiple instances of the same assay
### just keep the first instance encountered
smp <- sampleMap(prad)
expr <- assays(prad)
for (k in 1:length(expr)) {
  samps <- smp[which(smp$assay==names(expr)[k]),]
  notdup <- samps[which(!duplicated(samps$primary)),"colname"]
  #message(sprintf("%s: %i notdup", names(expr)[k], length(notdup)))
  prad[[k]] <- suppressMessages(prad[[k]][,notdup])
}


### create ID, STATUS columns, remove spaces/hyphens from patient labels
pID <- colData(prad)$patientID
colData(prad)$ID <- pID
colData(prad)$STATUS <- pam50
colData(prad)$STATUS <- gsub(" ",".",colData(prad)$STATUS)
colData(prad)$STATUS <- gsub("-",".",colData(prad)$STATUS)


#reference: Shraddha Pai
```

check if data was preprocessed 
```{r}
pheno2 <- colData(prad)
head(pheno2[,c("ID","STATUS")])
dim(pheno2)

table(pheno2$STATUS)
table(pheno2$STATUS,useNA="always")  # good practice: useNA="always" shows missing values
```


After data processsing PRAD is now a subset of prostate cancer cases AND primary cancer samples AND with responses for recurrence (yes,no) only
Creating subset of omics data for cohort of interest: 
```{r}
mir.subset <- assays(prad)[["PRAD_miRNASeqGene-20160128"]]


prot.subset <- assays(prad)[["PRAD_RPPAArray-20160128"]]

rna.subset <- assays(prad)[["PRAD_RNASeq2GeneNorm-20160128"]]


mut.subset <- assays(prad)[["PRAD_Mutation-20160128"]]

dim(mir.subset)



dim(prot.subset)

dim(rna.subset)

dim(mut.subset)
```

Note that NOT all assay are available for all the patients in our data i.e we have 395 patients in the PRAD dataset but have 393 Mir assays, we will consider this when merging columns with phenotype data and removing NAs



We will now subset Phenotype data to variables of our interest. There are 1147 variables, we will use random forest, domain knowledge and literature search to keep phenotype information predictive of recurrence

read in data
```{r}

dat <- data.frame(pheno2) #pheno2 (phenotype data) was created during data preprocessing step when subsetting the cohort

dim(dat)

names(dat)
```



merge MiRNA dataset with status and ID
```{r}


 
#transpose MiRNA data
tmir <- t(mir.subset)


# ID column [,1] does not have header
tmir[,1]


class(tmir)

# convert to dataframe to  add header
tmir1 <- as.data.frame(tmir)

colnames(tmir1)


str(tmir1)

#create patientID column to merge later
rownames(tmir1)

# Add header
tmir1$patientID = rownames(tmir1)

# recheck if patient id header was added
tmir1$patientID

```

Merge pheno and MiRNa data (left join by PatientID)
```{r}
 test <- as.data.frame(pheno2) %>% 
   select(c(patientID,STATUS))

#check str of IDs
str(test$patientID) 
str(tmir1$patientID)

```

```{r}
#substring ID to keep them consistent across datasets for merging 
tmir1$patientID <- substr(tmir1$patientID,start=1,stop=12)
```

Merge datasets
```{r}
combined.mir <- left_join(test, tmir1, by = "patientID")

```

MiRNA has 393 obs and combined has 395 obs, remove if a row has NAs across all columns, since miRNA data was not available for all the patients, will remove any NAs 
```{r}
combined.mir1 <- na.omit(combined.mir)

dim(combined.mir1)
```
Reformat and prepare data for Exploratory data analysis and ML models
```{r}
str(combined.mir1)

names(combined.mir1)

#change omic values to numeric, omic values are already numeric

#combined.mir1[,-c(1,2)] <- lapply(combined.mir1[,-c(1,2)],as.numeric)

#drop patinetID columns

combined.mir1 <- combined.mir1[,-1]

#change outcome to as.factor
combined.mir1$STATUS <- as.factor(combined.mir1$STATUS)


table(combined.mir1$STATUS)

str(combined.mir1)

```




